frontend/src/Components/ShippingInfo/ShippingInfo.jsx

```diff
@@ -1,85 +1,231 @@
+import { addBookSelector } from "../../Redux/selector";
+import { Link, useNavigate } from "react-router-dom";
+import { setShippingInfo } from "../../Redux/actions";
 
 function ShippingInfo() {
-    const dispatch = useDispatch();
-    const cartBooks = useSelector(addBookSelector);
+  const navigate = useNavigate();
+  const dispatch = useDispatch();
+  const cartBooks = useSelector(addBookSelector);
 
-    // State để lưu trữ thông tin vận chuyển
-    const [shippingDetails, setShippingDetails] = useState({
-        recipientName: '',
-        companyName: '',
-        streetAddress: '',
-        landmark: '',
-        country: 'VietNam',
-        cityName: '',
-        zipCode: '',
-        mobile: '',
-        phone: ''
+  // State to store shipping details and errors
+  const [shippingDetails, setShippingDetails] = useState({
+    recipientName: "",
+    companyName: "",
+    streetAddress: "",
+    landmark: "",
+    country: "VietNam",
+    cityName: "",
+    zipCode: "",
+    mobile: "",
+    phone: "",
+  });
+
+  const [errors, setErrors] = useState({
+    recipientName: false,
+    streetAddress: false,
+    mobile: false,
+    phone: false,
+  });
+
+  const [savedAddresses, setSavedAddresses] = useState([]);
+
+
+  useEffect(() => {
+    const saved = localStorage.getItem("savedAddresses");
+    if (saved) {
+      setSavedAddresses(JSON.parse(saved));
+    }
+  }, []);
+
+
+  const saveToLocalStorage = (address) => {
+    const updatedAddresses = [...savedAddresses, address];
+    setSavedAddresses(updatedAddresses);
+    localStorage.setItem("savedAddresses", JSON.stringify(updatedAddresses));
+  };
+
+  const totalQuantity = cartBooks.reduce((acc, book) => acc + book.quantity, 0);
+  const subTotal = cartBooks.reduce(
+    (acc, book) => acc + book.quantity * book.newPrice,
+    0
+  );
+
+  const handleInputChange = (e) => {
+    const { name, value } = e.target;
+    setShippingDetails({
+      ...shippingDetails,
+      [name]: value,
     });
+  };
 
-    // Tính tổng số lượng và tổng giá tiền
-    const totalQuantity = cartBooks.reduce((acc, book) => acc + book.quantity, 0);
-    const subTotal = cartBooks.reduce((acc, book) => acc + book.quantity * book.newPrice, 0);
+  const handleSaveAndContinue = () => {
+    if (validateForm()) {
+      saveToLocalStorage(shippingDetails);
+      dispatch(setShippingInfo(shippingDetails));
+      navigate("/confirm");
+    }
+  };
 
-    const handleInputChange = (e) => {
-        const { name, value } = e.target;
-        setShippingDetails({
-            ...shippingDetails,
-            [name]: value
-        });
+  // Validate form input
+  const validateForm = () => {
+    let isValid = true;
+    const newErrors = {
+      recipientName: false,
+      streetAddress: false,
+      mobile: false,
+      phone: false,
     };
+    if (!shippingDetails.recipientName) {
+      newErrors.recipientName = true;
+      isValid = false;
+    }
+    if (!shippingDetails.streetAddress) {
```

Comment:
For validation, should be handle by libs
https://www.react-hook-form.com/ or https://formik.org/ and https://www.npmjs.com/package/yup